File: ui/src/lib/api/leaderboard.ts
import api from "./api";

export const leaderboardApi = {
  get: (boardType: string, period: string, limit = 100) =>
    api(`/leaderboard/${boardType}/${period}?limit=${limit}`),

  myRank: (boardType: string, period: string) =>
    api(`/leaderboard/${boardType}/${period}/me`),

  aroundUser: (boardType: string, period: string, userId: string, count = 5) =>
    api(`/leaderboard/${boardType}/${period}/around/${userId}?count=${count}`),
};

File: app/use_cases/challenges/leaderboard.py
"""Leaderboard use cases — get rankings."""

from __future__ import annotations

import uuid

from app.interfaces.services.leaderboard_service import (ILeaderboardService,
                                                         LeaderboardEntry,
                                                         LeaderboardPeriod,
                                                         LeaderboardType)


class GetLeaderboardUseCase:
    def __init__(self, leaderboard_service: ILeaderboardService) -> None:
        self._leaderboard = leaderboard_service

    async def execute(
        self,
        board_type: LeaderboardType,
        period: LeaderboardPeriod,
        limit: int = 100,
    ) -> list[LeaderboardEntry]:
        return await self._leaderboard.get_top(board_type, period, limit=limit)


class GetUserRankUseCase:
    def __init__(self, leaderboard_service: ILeaderboardService) -> None:
        self._leaderboard = leaderboard_service

    async def execute(
        self,
        user_id: uuid.UUID,
        board_type: LeaderboardType,
        period: LeaderboardPeriod,
    ) -> LeaderboardEntry | None:
        return await self._leaderboard.get_rank(user_id, board_type, period)
